File: dsaWithJava/sorting/QuickSort.java
package sorting;

import java.util.Scanner;

/*
* quick sort = moves smaller elements to left of a pivot recursively * * divide array in 2 partitions
* run-time complexity = Best case O(n log(n))
*                       Average case O(n log(n))
*                       Worst case O(n^2) if already sorted
* space complexity = O(log(n)) due to recursion
*/
public class QuickSort {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int[] arr = SortHelper.takeInput();
        int sort = SortHelper.sort();
        quickSort(arr, sort);
        SortHelper.printSortedArray(arr);
        sc.close();
    }

    public static void quickSort(int[] arr, int sort) {
        quickSort(arr, 0, arr.length - 1, sort);
    }

    /*
     * The main function that implements QuickSort
     * arr[] --> Array to be sorted,
     * start --> Starting index,
     * end --> Ending index
     */
    private static void quickSort(int[] arr, int start, int end, int sort) {
        // base case
        if (start < end) {
            // partition fn will return the right position of the pivotIndex
            int pivotIndex = partition(arr, start, end, sort);
            // recursive call to sort smaller elements than pivot
            quickSort(arr, start, pivotIndex - 1, sort);
            // recursive call to sort greater elements than pivot
            quickSort(arr, pivotIndex + 1, end, sort);
        }
    }

    /*
     * This function takes last element as pivot, places
     * the pivot element at its correct position in sorted
     * array, and places all smaller (smaller than pivot)
     * to left of pivot and all greater elements to right
     * of pivot
     */
    private static int partition(int[] arr, int start, int end, int sort) {
        // last element taken as pivot
        int pivot = arr[end];
        // i & j are pointers
        // when element at j is smaller/greater than the pivot,
        // the elements at i & j index will be swapped
        int i = start - 1;
        for (int j = start; j < end; j++) {
            // For Sorting in Ascending Order
            if (sort == 1) {
                // If current element is smaller than the pivot
                if (arr[j] < pivot) {
                    // i is increased to make place for the smaller value
                    i++;
                    // swapping elements, this makes all elements smaller than the pivot,
                    // shifted to the left of the pivotIndex
                    swap(arr, i, j);
                }
            }
            // For Sorting in Descending Order
            else {
                // If current element is greater than the pivot
                if (arr[j] > pivot) {
                    // i is increased to make place for the greater value
                    i++;
                    // swapping elements, this makes all elements greater than the pivot,
                    // shifted to the left of the pivotIndex
                    swap(arr, i, j);
                }
            }
        }
        // after increasing i, this will be the final index of the pivot
        i++;
        // after transvering the array i.e. till the pivot, 
        // the pivot will be swapped with the value at i
        swap(arr, i, end);
        return i;
    }

    // Fn to swap elements
    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}

File: Rust/src/sorting/quick_sort.rs
pub use super::SortOrder;

/// Quick sort algorithm
///
/// # Arguments
/// * `arr` - A mutable reference to an array to be sorted.
/// * `order` - Sort order.
///
/// # Examples
/// ```
/// use rust::sorting::{quick_sort, SortOrder};
///
/// let mut arr = [1, 3, 2, 5, 4];
/// quick_sort(&mut arr, SortOrder::Asc);
/// assert_eq!(arr, [1, 2, 3, 4, 5]);
/// ```
///
/// # Time complexity
/// * O(n log n)
///
/// # Space complexity
/// * O(log n)
pub fn quick_sort<T: PartialOrd + Copy>(arr: &mut [T], order: SortOrder) {
    let len = arr.len();

    if len <= 1 {
        return;
    }

    let pivot = partition(arr, order);

    // recursive call to sort smaller elements than pivot
    quick_sort(&mut arr[..pivot], order);
    // recursive call to sort larger elements than pivot
    quick_sort(&mut arr[pivot + 1..], order);
}

// This function takes last element as pivot, places
// the pivot element at its correct position in sorted
// array, and places all smaller (smaller than pivot)
// to left of pivot and all greater elements to right
// of pivot
fn partition<T: PartialOrd + Copy>(arr: &mut [T], order: SortOrder) -> usize {
    let len = arr.len();
    let pivot_el = arr[len - 1];
    let mut i = 0; // index of smaller element

    // loop through all elements
    for j in 0..len - 1 {
        // match order and compare arr[j] and pivot_el
        if match order {
            SortOrder::Asc => arr[j] < pivot_el,
            SortOrder::Des => arr[j] > pivot_el,
        } {
            // if current element is smaller than the pivot
            // swap arr[i] and arr[j], this will move all
            // smaller elements to the left of pivot
            arr.swap(i, j);
            i += 1; // increment index of smaller element
        }
    }

    // after loop, swap pivot element with arr[i]
    arr.swap(i, len - 1);

    i
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn quick_sort_asc() {
        let mut arr = [1, 3, 2, 5, 4];
        quick_sort(&mut arr, SortOrder::Asc);
        assert_eq!(arr, [1, 2, 3, 4, 5]);
    }

    #[test]
    fn quick_sort_des() {
        let mut arr = [1, 3, 2, 5, 4];
        quick_sort(&mut arr, SortOrder::Des);
        assert_eq!(arr, [5, 4, 3, 2, 1]);
    }
}
